lib/ex_crypto.ex

```diff
@@ -211,18 +211,63 @@ defmodule ExCrypto do
   """
   @spec encrypt(binary, binary, binary, binary) :: {:ok, {binary, {binary, binary, binary}}} | {:error, binary}
   def encrypt(key, authentication_data, initialization_vector, clear_text) do
-    case :crypto.block_encrypt(:aes_gcm, key, initialization_vector, {authentication_data, clear_text}) do
-      {cipher_text, cipher_tag} -> {:ok, {authentication_data, {initialization_vector, cipher_text, cipher_tag}}}
+    _encrypt(key, initialization_vector, {authentication_data, clear_text}, :aes_gcm)
+  end
+
+
+  @doc """
+  Encrypt a `binary` with AES in CBC mode.
+
+  Returns a tuple containing the `initialization_vector`, and `cipher_text`.
+
+  At a high level encryption using AES in CBC mode looks like this:
+
+      key + clear_text -> init_vec + cipher_text
+
+  ## Examples
+
+      iex> clear_text = "my-clear-text"
+      iex> {:ok, aes_256_key} = ExCrypto.generate_aes_key(:aes_256, :bytes)
+      iex> {:ok, {iv, payload}} = ExCrypto.encrypt(aes_256_key, clear_text)
+      iex> {iv, cipher_text} = payload
+      iex> assert(is_bitstring(cipher_text))
+      true
+
+  """
+  @spec encrypt(binary, binary, binary) :: {:ok, {binary, {binary, binary}}} | {:error, binary}
+  def encrypt(key, clear_text) do
```

Comment:
Yes, you should always generate the HMAC on the cipher-text so that on the flip side you can do message authentication before running any decryption.

```
encrypt -> compute HMAC -> transmit -> verify HMAC -> decrypt
```

You should never decrypt before verifying the HMAC. That's why GCM is so nice, it handles it inside the algorithm and always in the correct order.

One thing we could do here is make make the "main" encrypt command (and it's decrypt counterpart) more of a canonical representation of how to call the helper methods, but not make the helper methods private.

That way if someone doesn't have an HMAC in a legacy app they can still decrypt in CBC mode by calling the helpers but skipping the HMAC steps.